don't write failure event to event log

jamiedemaria opened the pull request:
## Summary & Motivation
In https://github.com/dagster-io/internal/pull/14457 the write path is behind a feature flag. We should be consistent in OSS and not write the event. 

Most of this is for test refactoring and making it easier to test in internal with the feature flag

## How I Tested These Changes

jamiedemaria commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28670?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#28670** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28670?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28670?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#28506** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28506?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#28095** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28095?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#28568** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28568?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#28283** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28283?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#28093** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28093?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 2 other dependent PRs ([#28094](https://github.com/dagster-io/dagster/pull/28094) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28094?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#28258](https://github.com/dagster-io/dagster/pull/28258) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28258?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

gibsondan reviewed (changes requested):
I feel a little nervous about the amount of duplicated logic here across different storages - seems easy to miss one and let things slip through the cracks.

I think every store event call in OSS outside of tests goes through DagsterInstance.store_event or DagsterInstance.handle_new_event, what do you think about adding the check there instead before it ever enters the storage? Maybe it could check asset_records_have_last_planned_and_failed_materializations which I think effectively gates the behavior as well?

jamiedemaria commented:
Yeah fair enough. I didn't do a good enough look at the callsites to see that everything actually went through the instance. Added the check to `store_event` and `handle_new_event`.  Updated!

gibsondan commented on python_modules/dagster/dagster/_core/instance/__init__.py:
```diff
@@ -2487,6 +2487,12 @@ def get_handlers(self) -> Sequence[logging.Handler]:
 
     def store_event(self, event: "EventLogEntry") -> None:
+        if (
+            event.dagster_event is not None
+            and event.dagster_event.is_asset_failed_to_materialize
+            and not self._event_storage.asset_records_have_last_planned_and_failed_materializations
```
could make this a helper method

gibsondan reviewed (approved):
looks good. maybe not directly relevant here but note that in the cloud event upload path this doesn't actually go through instance.store_event (maybe it should) but rather skips straight to event_storage.store_event
